src/icon_dearchiver.cc

```diff
@@ -0,0 +1,480 @@
+    uncompressed_offsets[ICON_TYPE_icl8] = icon_offsets[2] - offset_base;
+    uncompressed_offsets[ICON_TYPE_icsN] = icon_offsets[3] - offset_base;
+    uncompressed_offsets[ICON_TYPE_ics4] = icon_offsets[4] - offset_base;
+    uncompressed_offsets[ICON_TYPE_ics8] = icon_offsets[5] - offset_base;
+  }
+  
+  strip_trailing_whitespace(icon_name);
+  icon_name = decode_mac_roman(icon_name, /*for filename=*/ true);
+  
+  write_icns(context, icon_number, icon_name, uncompressed_data.data(), uncompressed_offsets);
+  
+  // Done: continue right after the icon, skipping any possible padding after
+  // the icon's data
+  r.go(r_where + icon_size);
+}
+
+
+int main(int argc, const char** argv) {
+  try {
+    if (argc < 2) {
+      print_usage();
+      return 2;
+    }
+    
+    DearchiverContext context;
+    
+    for (int x = 1; x < argc; x++) {
+      if (context.base_name.empty()) {
+        context.base_name = argv[x];
+      } else if (context.out_dir.empty()) {
+        context.out_dir = argv[x];
+      } else {
+        fprintf(stderr, "excess argument: %s\n", argv[x]);
+        print_usage();
+        return 2;
+      }
+    }
+
+    if (context.base_name.empty()) {
+      print_usage();
+      return 2;
+    }
+    if (context.out_dir.empty()) {
+      context.out_dir = string_printf("%s.out", context.base_name.c_str());
+    }
+    mkdir(context.out_dir.c_str(), 0777);
+    
+    string        content = load_file(context.base_name);
+    StringReader& r = context.in = StringReader(content);
+
+    // Check signature ('QBSE' 'PACK')
+    if (r.get_u32b() != 0x51425345 || r.get_u32b() != 0x5041434B) {
+      fprintf(stderr, "File '%s' isn't an Icon Archiver file\n", context.base_name.c_str());
+      return 2;
+    }
+    
+    // ???
+    r.skip(2);
+    
+    // Version: 1 = Icon Archiver 2; 2 = Icon Archiver 4
+    uint16_t version = r.get_u16b();
+    if (version != 1 && version != 2) {
+      fprintf(stderr, "File '%s' has unsupported version %hu\n", context.base_name.c_str(), version);
+      return 2;
+    }
+    
+    uint32_t icon_count = r.get_u32b();
+    
+    // ???
+    r.skip(32);
+    
+    // Seems to be some kind of date
+    r.get_u64b();
+    
+    // ???
+    r.skip(8);
+    
+    if (version > 1) {
+      // Another signature? ('IAUB')
+      if (r.get_u32b() != 0x49415542) {
+        fprintf(stderr, "File '%s' isn't an Icon Archiver version 2 file\n", context.base_name.c_str());
+        return 2;
+      }
+      
+      // ???
+      r.skip(57);
+      
+      // Are the copyright and comment string locked, i.e. can't be changed
+      // anymore in Icon Archiver
+      r.get_u8();
+      
+      // ???
+      r.skip(2);
+      
+      // Copyright and comment strings are Pascal strings padded to a fixed length
+      r.get_u8();
+      string copyright = decode_mac_roman(r.readx(63));
+      
+      r.get_u8();
+      string comment = decode_mac_roman(r.readx(255));
```

Comment:
These might need to have strip_trailing_zeroes called on them - it looks like decode_mac_roman will just copy over the zeroes if present.

Reply:
Fixed.